homeassistant/components/asuswrt/config_flow.py

```diff
@@ -0,0 +1,175 @@
+"""Config flow to configure the AsusWrt integration."""
+import logging
+import os
+import socket
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_MODE,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_PROTOCOL,
+    CONF_USERNAME,
+)
+from homeassistant.helpers import config_validation as cv
+
+# pylint:disable=unused-import
+from .const import (
+    CONF_DNSMASQ,
+    CONF_INTERFACE,
+    CONF_REQUIRE_IP,
+    CONF_SSH_KEY,
+    DEFAULT_DNSMASQ,
+    DEFAULT_INTERFACE,
+    DEFAULT_SSH_PORT,
+    DOMAIN,
+)
+from .router import get_api
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _is_file(value) -> bool:
+    """Validate that the value is an existing file."""
+    file_in = os.path.expanduser(str(value))
+
+    if not os.path.isfile(file_in):
+        return False
+    if not os.access(file_in, os.R_OK):
+        return False
+    return True
+
+
+def _get_ip(host):
+    """Get the ip address from the host name."""
+    if host is None:
+        return None
+    try:
+        return socket.gethostbyname(host)
+    except socket.gaierror:
+        return None
+
+
+class AsusWrtFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize AsusWrt config flow."""
+        self._host = None
+        self._name = None
+
+    def _show_setup_form(self, user_input=None, errors=None):
+        """Show the setup form to the user."""
+
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                    vol.Optional(CONF_NAME, default=user_input.get(CONF_NAME, "")): str,
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Optional(CONF_PASSWORD): str,
+                    vol.Optional(CONF_SSH_KEY): str,
+                    vol.Required(CONF_PROTOCOL, default="ssh"): vol.In(
+                        {"ssh": "SSH", "telnet": "Telnet"}
+                    ),
+                    vol.Required(CONF_PORT, default=DEFAULT_SSH_PORT): cv.port,
+                    vol.Required(CONF_MODE, default="router"): vol.In(
+                        {"router": "Router", "ap": "Access Point"}
+                    ),
+                    vol.Required(CONF_REQUIRE_IP, default=True): bool,
```

Comment:
Is this option required to be able to connect to the router? We only want required options for the connection in the main flow. Optional options should go in an options flow.

Reply:
All this parameters are used by API constructor, cannot be changed after router instance is created and cannot be changed at runtime.

Reply:
Right, but can we connect to the router regardless of this option? We can reload the integration when changing this option in an options flow.

Reply:
In this case same consideration can be done for next 2 config option. But why don't allow the user to select the right option during setup? And how I reload integration (some example)?

Reply:
As I said we only want the required settings in the main flow.

https://developers.home-assistant.io/docs/config_entries_options_flow_handler#signal-updates

Reload the config entry with `hass.config_entries.async_reload`.